src/Fl_Anim_GIF_Image.cxx

```diff
@@ -0,0 +1,1279 @@
+      DEBUG(("loop count %d reached - stopped!\n", fi_->loop_count));
+      stop();
+    }
+    else
+      frame = 0;
+  }
+  if (frame >= fi_->frames_size)
+    return false;
+  set_frame(frame);
+  double delay = fi_->frames[frame].delay;
+  if (min_delay && delay < min_delay) {
+    DEBUG(("#%d: correct delay %f => %f\n", frame, delay, min_delay));
+    delay = min_delay;
+  }
+  if (is_animated() && delay > 0 && speed_ > 0) {  // normal GIF has no delay
+    delay /= speed_;
+    Fl::add_timeout(delay, cb_animate, this);
+  }
+  return true;
+}
+
+
+/*virtual*/
+void Fl_Anim_GIF_Image::on_frame_data(Fl_GIF_Image::GIF_FRAME &gf) {
+  fi_->on_frame_data(gf);
+}
+
+
+/*virtual*/
+void Fl_Anim_GIF_Image::on_extension_data(Fl_GIF_Image::GIF_FRAME &gf) {
+  fi_->on_extension_data(gf);
+}
+
+
+/** Resizes the image to the specified size, replacing the current image.
+
+ If \ref DONT_RESIZE_CANVAS is not set, the canvas widget will also be resized.
+
+ \param[in] w, h new size of teh naimtion frames
+ */
+Fl_Anim_GIF_Image& Fl_Anim_GIF_Image::resize(int w, int h) {
+  int W(w);
+  int H(h);
+  if (canvas_ && !W && !H) {
+    W = canvas_->w();
+    H = canvas_->h();
+  }
+  if (!W || !H || ((W == this->w() && H == this->h()))) {
+    return *this;
+  }
+  fi_->resize(W, H);
+  scale_frame(); // scale current frame now
+  this->w(fi_->canvas_w);
+  this->h(fi_->canvas_h);
+  if (canvas_ && !(flags_ & DONT_RESIZE_CANVAS)) {
+    canvas_->size(this->w(), this->h());
+  }
+  return *this;
+}
+
+
+/** Resizes the image to the specified size, replacing the current image.
+
+ If \ref DONT_RESIZE_CANVAS is not set, the canvas widget will also be resized.
+
+ \param[in] scale rescale factor in relation to current size
+ */
+Fl_Anim_GIF_Image& Fl_Anim_GIF_Image::resize(double scale) {
+  return resize((int)lround((double)w() * scale), (int)lround((double)h() * scale));
+}
+
+
+void Fl_Anim_GIF_Image::scale_frame() {
+  int i(frame_);
+  if (i < 0)
+    return;
+  fi_->scale_frame(i);
+}
+
+
+void Fl_Anim_GIF_Image::set_frame() {
+  int i(frame_);
+  if (i < 0)
+    return;
+  fi_->set_frame(i);
+}
+
+
+void Fl_Anim_GIF_Image::set_frame(int frame) {
+//  int last_frame = frame_;
+  frame_ = frame;
+  // NOTE: uncaching decreases performance, but saves a lot of memory
+  if (uncache_ && this->image())
+    this->image()->uncache();
+
+  fi_->set_frame(frame_);
+
+  Fl_Widget* cv = canvas();
+  if (cv) {
+    Fl_Group* parent = cv->parent();
+    bool no_bg = (cv->box() == FL_NO_BOX);
+    bool outside = (!(cv->align() & FL_ALIGN_INSIDE) && !((cv->align() & FL_ALIGN_POSITION_MASK)==FL_ALIGN_CENTER));
+//    bool dispose =    (fi_->frames[last_frame].dispose == FrameInfo::DISPOSE_BACKGROUND)
+//                   || (fi_->frames[last_frame].dispose == FrameInfo::DISPOSE_PREVIOUS);
+    if (parent && (no_bg || outside))
+      parent->redraw();
+    else
+      cv->redraw();
+
+// Note: the code below did not animate labels with a pixmap outside of the canvas
```

Comment:
@wcout  I did simplify the code below to make things work correctly in `test/pixmap`, but I am not entirely sure if I messed up something that needs to be there. Maybe I am drawing the part image too often now?

Reply:
I will have a look...

Reply:
> @wcout I did simplify the code below to make things work correctly in `test/pixmap`, but I am not entirely sure if I messed up something that needs to be there. Maybe I am drawing the part image too often now?

**Short answer:**
Change is good. Please use it.

**Long answer:**
I tested your change with different GIF's. Everything worked, and probably even better (less redraws). Looking at my commit history I think the code in this method was "ported" from an (working) predecessor version that had another approach and was then never tested with outside labels.

BTW: Thank you for all your effort. I wouldn't have expected it to be so much work...

Reply:
Ok, thanks for checking my change.

Thanks for implementing this in the first place. It's a great piece of code. Most of my changes were cosmetic to fit the documentation style and to not overload the test folder. I know that if I don't do that before the merge, we will never touch the code again ;-) .